src/SizedArray.jl

```diff
@@ -16,6 +16,14 @@ struct SizedArray{S <: Tuple, T, N, M} <: StaticArray{S, T, N}
         if length(a) != tuple_prod(S)
             error("Dimensions $(size(a)) don't match static size $S")
         end
+        if size(a) != size_to_tuple(S)
+            Base.depwarn("Construction of `SizedArray` with an `Array` of a different
+                size is deprecated. If you need this functionality report it at
+                https://github.com/JuliaArrays/StaticArrays.jl/pull/666 .
+                Calling `sa = reshape(a::Array, s::Size)` will actually reshape
+                array `a` in the future and converting `sa` back to `Array` will
+                return an `Array` of shape `s`.", :SizedArray)
```

Comment:
Oh sorry! This isn't quite what I meant. I meant that we should make this depwarn in the conversions and constructors (just a tweak to what you had already).

Having said that... this is an clearer alternative to resolve these issues by clarifying the design rather than hacking around the `SizedArray`-might-also-reshape design that we have already.

I do think it might be better.  Any thoughts @andyferris?

Reply:
Yes, I think keeping the old behavior + depwarn for the next version and then changing it in the following one to something reasonable would be least surprising.